rclrs/src/dynamic_message/message_structure.rs

```diff
@@ -0,0 +1,254 @@
+    },
+    Message(Box<MessageStructure>),
+}
+
+/// A description of a single field in a [`DynamicMessage`][1].
+///
+/// The concrete type of a field is the combination of its [`BaseType`] with its [`ValueKind`].
+/// That is, the base types exist as single values, arrays, bounded sequences and unbounded sequences.
+///
+/// [1]: crate::dynamic_message::DynamicMessage
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MessageFieldInfo {
+    /// The field name.
+    pub name: String,
+    /// The base type – number, string, etc.
+    pub base_type: BaseType,
+    /// Whether the field is a simple value, an array, or a (bounded) sequence.
+    pub value_kind: ValueKind,
+    pub(crate) string_upper_bound: usize,
+    pub(crate) resize_function:
+        Option<unsafe extern "C" fn(arg1: *mut std::os::raw::c_void, size: usize) -> bool>,
+    pub(crate) offset: usize,
+}
+
+/// A description of the structure of a message.
+///
+/// Namely, the list of fields and their types.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MessageStructure {
+    /// The set of fields in the message, ordered by their offset in the message.
+    ///
+    /// A `Vec` is easier to handle and faster than a `HashMap` for typical numbers of message fields.
+    /// If you need a `HashMap`, simply create your own from this `Vec`.
+    pub fields: Vec<MessageFieldInfo>,
+    /// The size of this structure in bytes.
+    pub size: usize,
+    /// The namespace of this type. This is something like `geometry_msgs__msg`.
+    pub namespace: String,
+    /// The name of this type. This does not contain the package name.
+    pub type_name: String,
+}
+
+/// Information on whether a field is a single value or a list of some kind.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ValueKind {
+    /// This field is a single value, which includes string values.
+    Simple,
+    /// This field is an array of values.
+    Array {
+        /// The array length.
+        length: usize,
+    },
+    /// This field is a [`Sequence`][1] of values.
+    ///
+    /// [1]: rosidl_runtime_rs::Sequence
+    Sequence,
+    /// This field is a [`BoundedSequence`][1] of values.
+    ///
+    /// [1]: rosidl_runtime_rs::BoundedSequence
+    BoundedSequence {
+        /// The maximum sequence length.
+        upper_bound: usize,
+    },
+}
+
+// ========================= impl for BaseType =========================
+
+impl BaseType {
+    // The inner message type support will be nullptr except for the case of a nested message.
+    // That function must be unsafe, since it is possible to safely create a garbage non-null
+    // pointer.
+    unsafe fn new(
+        type_id: u8,
+        string_upper_bound: Option<NonZeroUsize>,
+        inner: *const rosidl_message_type_support_t,
+    ) -> Self {
+        #[cfg(all(not(ros_distro = "foxy"), not(ros_distro = "galactic")))]
+        use rosidl_typesupport_introspection_c_field_types::*;
+        match u32::from(type_id) {
```

Comment:
Wouldn't this look cleaner with a lookup table?

Reply:
Hmm probably! I'll try it out.

Reply:
Two issues with that:
* The lookup table would be brittle, because the order of entries has to match the order of variants in the enum, but there's no clean way to check that it does without incurring overhead afaik
* The string, sequence and message cases are not a pure lookup, but have logic in their match arms. So the lookup table would have to exclude those and handle them specially, which would make this function look less clean again.

So I'd leave it as it is.

Reply:
Fair enough. The code as it is is not the prettiest, but does the job.